### PR TITLE
RegexArrayShapeMatcher - Fix PREG_UNMATCHED_AS_NULL with optional leading groups

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/11331

### CODE REVIEW

**Comment by staabm** [comment]
//cc @Seldaek

File: tests/PHPStan/Analyser/nsrt/bug-11311.php [hunk 1]
```diff
@@ -29,3 +29,40 @@ function unmatchedAsNullWithOptionalGroup(string $s): void {
...
+}
+
+function bug11331c(string $url):void {
+	if (preg_match('{^
+	(?:
+		(?:https?|git)://([^/]+)/ (?# group 1 here can be null if group 2 matches)
+		|                         (?# the alternation making it so that only either should match)
+		git@([^:]+):/?            (?# group 2 here can be null if group 1 matches)
+	)
+	(?# removing what follows makes the two first groups nullable, although it then has group 2 unsettable which looks buggy too as PREG_UNMATCHED_AS_NULL is present)
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/bug-11311.php]: @Seldaek not sure what this means. could you come up with a concrete code sample and assertions for it?

(sounds unrelated to the other bugs and might be worth a separate report)

Comment (Seldaek) [reply tests/PHPStan/Analyser/nsrt/bug-11311.php]: I just meant if you delete the lines of regex below this comment, then the type of the groups 1 and 2 become correct.

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/bug-11311.php]: I see thanks. I will remove the comment from the test so it not confuses others :)

**Review by Seldaek — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**